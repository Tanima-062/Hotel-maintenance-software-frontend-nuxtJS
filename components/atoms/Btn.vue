<template>
  <button
    :type="type"
    :tabindex="tabindex"
    class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out"
    :class="styleColor"
    @click="clickEvent"
    @mouseover="mouseover"
    @mouseleave="mouseleave"
  >
    <slot />
  </button>
</template>
<script lang="ts">
import { Prop, Vue, Component, Emit } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: 'button' })
  type!: string

  @Prop({ default: false })
  outline!: boolean

  @Prop({ default: 'blue' })
  color!: string

  @Prop({ default: 'black' })
  textColor!: string

  @Prop({ default: '' })
  height!: string

  @Prop({ default: false })
  isActive!: boolean

  @Prop({ default: false })
  isDisabled!: boolean

  @Prop({ default: false })
  isNotFocusLine!: boolean

  private isAnim: boolean = false
  private tabindex: number = 0

  get styleColor(): string {
    const classList = []

    if (this.outline) {
      classList.push('border border-solid')
    }

    // hoverとcursor
    this.switchHoverAndCusor(classList)

    // 文字と背景色の切り替え
    this.switchColor(classList)

    // 高さ設定
    this.switchHeight(classList)

    return classList.join(' ')
  }

  get isPushed(): boolean {
    return this.isAnim || this.isActive
  }

  private switchHoverAndCusor(classList: string[]): void {
    if (this.isDisabled) {
      classList.push('cursor-not-allowed ')
      this.tabindex = -1
    } else if (this.isPushed) {
      classList.push('cursor-default')
    } else if (!this.isNotFocusLine) {
      classList.push('focus:shadow-outline')
    }
  }

  private switchColor(classList: string[]): void {
    switch (true) {
      case !this.outline && this.color === 'blue':
        // ボタンdisabled時
        if (this.isDisabled) {
          classList.push('bg-btnColor-default__disabled text-white')
        } else {
          classList.push(
            'bg-btnColor text-white hover:bg-btnColor-default__hover'
          )
        }
        break
      case !this.outline && this.color === 'red':
        classList.push('text-white')
        classList.push('bg-btnColor-red hover:bg-opacity-90')
        break
      case !this.outline && this.color === 'green':
        classList.push('text-white')
        if (this.isDisabled) {
          classList.push('bg-btnColor-border')
          break
        }
        classList.push('bg-btnColor-green hover:bg-opacity-90')
        break
      case !this.outline && this.color === 'gray':
        if (this.isDisabled) {
          classList.push('bg-btnColor-cancel__disabled text-black-light')
        } else {
          classList.push('text-white')
          classList.push('bg-btnColor-cancel hover:bg-btnColor-cancel__hover')
        }
        break
      case !this.outline && this.color === 'white':
        classList.push('text-black ')
        classList.push(
          this.isPushed
            ? 'bg-white border border-btnColor-cancel'
            : 'bg-white border border-btnColor-cancel hover:bg-opacity-80'
        )
        break
      case this.outline && this.color === 'blue':
        // ボタンdisabled時
        if (this.isDisabled) {
          classList.push('border-btnColor text-btnColor')
          break
        }
        // ボタン通常時
        classList.push(
          'border-btnColor-outline text-btnColor-outline hover:border-btnColor-outline__hover hover:text-btnColor-outline__hover'
        )
        break
      case this.outline && this.color === 'gray':
        if (this.isDisabled) {
          // TODO: active時の色不明の為、決まり次第実装
          classList.push('bg-black border-black text-bdColor')
          break
        }

        classList.push('border-btnColor-cancel text-btnColor-cancel')

        break
      case this.color === 'none':
        if (!this.isDisabled) {
          classList.push(`${this.textColor} hover:${this.textColor}__hover`)
        } else {
          classList.push(`text-btnColor-disabled`)
        }
        break
      default:
        break
    }
  }

  private switchHeight(classList: string[]): void {
    if (this.height === 'small') {
      classList.push('h-8')
    } else if (this.height === 'big') {
      classList.push('h-11')
    } else if (this.height && this.height.indexOf('py-') === 0) {
      classList.push(this.height)
    } else {
      classList.push('py-3')
    }
  }

  @Emit('click')
  private clickEvent(e: Event): Event | void {
    if (this.isDisabled) return

    return e
  }

  @Emit()
  mouseover(e: Event): Event {
    return e
  }

  @Emit()
  mouseleave(e: Event): Event {
    return e
  }
}
</script>
<style lang="sass" scoped>
.btn
  transition: all .3s ease
</style>
