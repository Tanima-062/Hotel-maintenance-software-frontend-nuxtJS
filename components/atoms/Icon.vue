<template>
  <div :class="styles">
    <svg
      class="w-7 h-7 fill-current"
      :class="`${color}`"
      :style="`width: ${iconWidth}px; height: ${iconHeight}px`"
      :viewBox="`0 0 24 24`"
    >
      <path :d="icon" />
    </svg>
  </div>
</template>
<script lang="ts">
import { Vue, Component, Prop } from 'nuxt-property-decorator'

@Component
export default class extends Vue {
  @Prop({ default: '' })
  icon!: string

  @Prop({ default: 'text-white' })
  color!: string

  @Prop({ default: '' })
  bgColor!: string

  @Prop({ default: false })
  round!: boolean

  @Prop({ default: 24 })
  iconWidth!: number

  @Prop({ default: 24 })
  iconHeight!: number

  @Prop({ default: '' })
  extraClass!: string

  get styles(): string {
    const classList = []
    if (this.bgColor) {
      classList.push(`${this.bgColor}`)
    }
    if (this.round) {
      classList.push('rounded-full')
    }
    if (this.extraClass) {
      classList.push(this.extraClass)
    }
    return classList.join(' ')
  }
}
</script>
