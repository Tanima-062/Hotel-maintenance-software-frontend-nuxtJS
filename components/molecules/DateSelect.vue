<template>
  <div class="flex justify-between items-center mb-2">
    <div class="flex">
      <Btn
        color="none"
        text-color="text-btnColor"
        class="items-center select-none flex flex-row mr-4"
        :is-disabled="isBeforeToday"
        :is-not-focus-line="true"
        @click="selectDate(-14)"
      >
        <Icon
          :icon="icon.doubleLeft"
          :color="`${
            isBeforeToday ? 'text-btnColor-disabled' : 'text-btnColor'
          }`"
        />
        {{ $t('calendar.pagination.twoWeekLeft') }}
      </Btn>
      <Btn
        color="none"
        text-color="text-btnColor"
        class="items-center select-none flex flex-row mr-4"
        :is-disabled="isBeforeToday"
        :is-not-focus-line="true"
        @click="selectDate(-1)"
      >
        <Icon
          :icon="icon.left"
          :color="`${
            isBeforeToday ? 'text-btnColor-disabled' : 'text-btnColor'
          }`"
        />
        {{ $t('calendar.pagination.leftButton') }}
      </Btn>
    </div>
    <div class="flex items-center">
      <Btn
        class="w-18 h-9 mr-2"
        :outline="true"
        color="blue"
        @click="checkAndChangeDate(today, true, true)"
      >
        {{ $t('calendar.searchDate.todayButton') }}
      </Btn>
      <date-picker
        ref="dateheader"
        v-model="dateValue"
        :value="dateSelected"
        format="YYYY/MM/DD"
        :disabled-date="isNotWithInAYear"
        type="date"
        :placeholder="$t('calendar.searchDate.selectDate')"
        @input="dateValue = $event"
      />

      <Btn
        class="w-18 h-9 ml-2"
        color="blue"
        @click="checkAndChangeDate(dateValue, false, true)"
      >
        {{ $t('calendar.searchDate.serachButton') }}
      </Btn>
    </div>
    <div class="flex">
      <Btn
        color="none"
        text-color="text-btnColor"
        class="items-center select-none flex flex-row mr-4"
        :is-disabled="isOneYearAfter"
        :is-not-focus-line="true"
        @click="selectDate(1)"
      >
        {{ $t('calendar.pagination.rightButton') }}
        <Icon
          :icon="icon.right"
          :color="`${
            isOneYearAfter ? 'text-pagingColor-disable' : 'text-btnColor'
          }`"
        />
      </Btn>
      <Btn
        color="none"
        text-color="text-btnColor"
        class="items-center select-none flex flex-row mr-4"
        :is-disabled="isOneYearAfter"
        :is-not-focus-line="true"
        @click="selectDate(14)"
      >
        {{ $t('calendar.pagination.twoWeekRight') }}
        <Icon
          :icon="icon.doubleRight"
          :color="`${
            isOneYearAfter ? 'text-pagingColor-disable' : 'text-btnColor'
          }`"
        />
      </Btn>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Emit } from 'vue-property-decorator'
import {
  mdiChevronDoubleLeft,
  mdiChevronLeft,
  mdiChevronRight,
  mdiChevronDoubleRight,
} from '@mdi/js'
import {
  isWithinInterval,
  startOfDay,
  endOfDay,
  addYears,
  addDays,
  isBefore,
  isAfter,
} from 'date-fns'

@Component({
  components: {
    Btn: () => import('~/components/atoms/Btn.vue'),
    Icon: () => import('~/components/atoms/Icon.vue'),
  },
})
export default class extends Vue {
  @Prop({ default: new Date() })
  dateSelected!: Date

  @Prop({
    default: () => true,
  })
  checkBeforeChange!: Function

  private dateValue: Date = new Date()
  private today: Date = addDays(new Date(), 0)
  private icon: { [key: string]: string } = {
    doubleLeft: mdiChevronDoubleLeft,
    left: mdiChevronLeft,
    right: mdiChevronRight,
    doubleRight: mdiChevronDoubleRight,
  }

  private isNotWithInAYear(date: Date): boolean {
    return !isWithinInterval(date, {
      start: startOfDay(this.today),
      end: startOfDay(addYears(this.today, 1)),
    })
  }

  get isBeforeToday(): boolean {
    return isBefore(startOfDay(this.dateSelected), endOfDay(this.today))
  }

  get isOneYearAfter(): boolean {
    return isAfter(
      endOfDay(this.dateSelected),
      startOfDay(addYears(this.today, 1))
    )
  }

  private async selectDate(daysToChange: number): Promise<void> {
    const isCheckedBefore = await this.checkBeforeChange()
    if (!isCheckedBefore) return
    if (this.isBeforeToday && this.isOneYearAfter) return

    const changedDate = addDays(this.dateSelected, daysToChange)
    const oneYearLater = addYears(this.today, 1)

    if (isAfter(endOfDay(changedDate), startOfDay(oneYearLater))) {
      this.dateValue = oneYearLater
      this.change(oneYearLater)
      return
    }
    if (isBefore(startOfDay(changedDate), endOfDay(this.today))) {
      this.dateValue = startOfDay(this.today)
      this.change(startOfDay(this.today))
      return
    }
    this.dateValue = startOfDay(changedDate)
    this.change(startOfDay(changedDate))
  }

  private async checkAndChangeDate(
    changeDate: Date,
    isToday: boolean = false,
    isNeedCheckBefore: boolean = false
  ): Promise<void> {
    if (isNeedCheckBefore) {
      const isCheckedBefore = await this.checkBeforeChange()
      if (!isCheckedBefore) return
    }
    this.change(changeDate, isToday)
  }

  @Emit()
  private change(changeDate: Date, isToday: boolean = false): Date {
    if (isToday || !changeDate) {
      this.dateValue = new Date()
      changeDate = this.dateValue
    }
    return changeDate
  }
}
</script>
