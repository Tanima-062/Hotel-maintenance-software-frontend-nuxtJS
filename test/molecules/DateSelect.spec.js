import { mount } from '@vue/test-utils'
import Vue from 'vue'
import VueI18n from 'vue-i18n'
import DatePicker from 'vue2-datepicker'
import DateSelect from '@/components/molecules/DateSelect.vue'
import Btn from '@/components/atoms/Btn.vue'
import Icon from '@/components/atoms/Icon.vue'

Vue.use(VueI18n)
Vue.use(DatePicker)
describe('Molecules DateSelect Check', () => {
  const props = {
    dateSelected: new Date('2021/01/01'),
  }
  test('DateSelect Test', () => {
    const wrapper = mount(DateSelect, {
      propsData: props,
      mocks: {
        $t: (msg) => msg,
      },
      stubs: {
        Btn,
        Icon,
      },
    })
    // check if it is a vue element
    expect(wrapper.vm).toBeTruthy()

    // check Snapshot
    expect(wrapper.element).toMatchSnapshot()
  })
})
