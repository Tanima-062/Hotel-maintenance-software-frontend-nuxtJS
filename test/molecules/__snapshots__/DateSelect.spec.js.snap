// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules DateSelect Check DateSelect Test 1`] = `
<div
  class="flex justify-between items-center mb-2"
>
  <div
    class="flex"
  >
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out items-center select-none flex flex-row mr-4 cursor-not-allowed  text-btnColor-disabled py-3"
      tabindex="-1"
      type="button"
    >
      <div
        class=""
      >
        <svg
          class="w-7 h-7 fill-current text-btnColor-disabled"
          style="width: 24px; height: 24px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M18.41,7.41L17,6L11,12L17,18L18.41,16.59L13.83,12L18.41,7.41M12.41,7.41L11,6L5,12L11,18L12.41,16.59L7.83,12L12.41,7.41Z"
          />
        </svg>
      </div>
      
      calendar.pagination.twoWeekLeft
    
    </button>
     
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out items-center select-none flex flex-row mr-4 cursor-not-allowed  text-btnColor-disabled py-3"
      tabindex="-1"
      type="button"
    >
      <div
        class=""
      >
        <svg
          class="w-7 h-7 fill-current text-btnColor-disabled"
          style="width: 24px; height: 24px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
          />
        </svg>
      </div>
      
      calendar.pagination.leftButton
    
    </button>
  </div>
   
  <div
    class="flex items-center"
  >
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out w-18 h-9 mr-2 border border-solid focus:shadow-outline border-btnColor-outline text-btnColor-outline hover:border-btnColor-outline__hover hover:text-btnColor-outline__hover py-3"
      tabindex="0"
      type="button"
    >
      
      calendar.searchDate.todayButton
    
    </button>
     
    <div
      class="mx-datepicker"
    >
      <div
        class="mx-input-wrapper"
      >
        <input
          autocomplete="off"
          class="mx-input"
          name="date"
          placeholder="calendar.searchDate.selectDate"
          type="text"
        />
        <i
          class="mx-icon-clear"
        >
          <svg
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M810.005333 274.005333l-237.994667 237.994667 237.994667 237.994667-60.010667 60.010667-237.994667-237.994667-237.994667 237.994667-60.010667-60.010667 237.994667-237.994667-237.994667-237.994667 60.010667-60.010667 237.994667 237.994667 237.994667-237.994667z"
            />
          </svg>
        </i>
        <i
          class="mx-icon-calendar"
        >
          <svg
            height="1em"
            viewBox="0 0 1024 1024"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M940.218182 107.054545h-209.454546V46.545455h-65.163636v60.50909H363.054545V46.545455H297.890909v60.50909H83.781818c-18.618182 0-32.581818 13.963636-32.581818 32.581819v805.236363c0 18.618182 13.963636 32.581818 32.581818 32.581818h861.090909c18.618182 0 32.581818-13.963636 32.581818-32.581818V139.636364c-4.654545-18.618182-18.618182-32.581818-37.236363-32.581819zM297.890909 172.218182V232.727273h65.163636V172.218182h307.2V232.727273h65.163637V172.218182h176.872727v204.8H116.363636V172.218182h181.527273zM116.363636 912.290909V442.181818h795.927273v470.109091H116.363636z"
            />
          </svg>
        </i>
      </div>
    </div>
     
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out w-18 h-9 ml-2 focus:shadow-outline bg-btnColor text-white hover:bg-btnColor-default__hover py-3"
      tabindex="0"
      type="button"
    >
      
      calendar.searchDate.serachButton
    
    </button>
  </div>
   
  <div
    class="flex"
  >
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out items-center select-none flex flex-row mr-4 text-btnColor hover:text-btnColor__hover py-3"
      tabindex="0"
      type="button"
    >
      
      calendar.pagination.rightButton
      
      <div
        class=""
      >
        <svg
          class="w-7 h-7 fill-current text-btnColor"
          style="width: 24px; height: 24px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
          />
        </svg>
      </div>
    </button>
     
    <button
      class="flex justify-center items-center leading-5 outline-none focus:outline-none rounded-md transition-all duration-300 ease-in-out items-center select-none flex flex-row mr-4 text-btnColor hover:text-btnColor__hover py-3"
      tabindex="0"
      type="button"
    >
      
      calendar.pagination.twoWeekRight
      
      <div
        class=""
      >
        <svg
          class="w-7 h-7 fill-current text-btnColor"
          style="width: 24px; height: 24px;"
          viewBox="0 0 24 24"
        >
          <path
            d="M5.59,7.41L7,6L13,12L7,18L5.59,16.59L10.17,12L5.59,7.41M11.59,7.41L13,6L19,12L13,18L11.59,16.59L16.17,12L11.59,7.41Z"
          />
        </svg>
      </div>
    </button>
  </div>
</div>
`;
